lib/tapioca/dsl/compilers/rspec.rb

```diff
@@ -54,13 +71,400 @@ def create_includes(klass)
+              possible = context.children.filter { |c| c.location.line >= rsig_node.location.line }
+              idx = possible.find_index { |n| n.rsigs.include?(rsig_node) }
+              return unless idx
+
+              line_number_key = T.let(nil, T.nilable(Integer))
+              target = T.let(nil, T.nilable(SemanticNode::SendNode))
+              possible[idx + 1..]&.each do |n|
+                sends = n.sends_no_receivers
+                sends.each do |send|
+                  next unless RSIG_TARGETS.include?(send.method_name)
+
+                  target = send
+                  line_number_key = send.location.line
+                  break
+                end
+                break if target
+
+                # Safety measure to make sure we don't match with an unexpected second sig
+                break if n.rsigs.any?
+              end
+              return unless target && line_number_key
+
+              [line_number_key, target]
+            end
+
+            sig {
+              params(mod: Module, path: String, ast: Parser::AST::Node,
+                rsigs_procs: T::Hash[Integer, [Thread::Backtrace::Location, Proc]])
+                .returns(T.attached_class)
+            }
+            def make(mod:, path:, ast:, rsigs_procs:)
+              root = SemanticNode.make(ast:)
+              rsig_nodes = index_by(root.rsigs) { |n| n.location.begin.line }
+              methods_by_line = rsigs_procs.map do |sig_line_number, (sig_caller_location, decl_proc)|
+                target_sig_node = rsig_nodes[sig_line_number]
+                next unless target_sig_node
+
+                rsig_target = find_rsig_target(target_sig_node)
+                next unless rsig_target
+
+                target_line_number, target_node = rsig_target
+
+                rsig_target_generated_method_name =
+                  (target_node.method_name == :subject) ? :subject : target_node.sym_args[0]
+                next unless rsig_target_generated_method_name
+
+                method_semantics = MethodSemantics.new(
+                  method_name: rsig_target_generated_method_name,
+                  target_node:,
+                  target_sig_node:,
+                  target_line_number:,
+                  sig_line_number:,
+                  sig_caller_location:,
+                  decl_proc:
+                )
+                [target_line_number, method_semantics]
+              end.compact.to_h
+
+              new(mod:, path:, root:, methods_by_line:)
+            end
+          end
+        end
+
+        class ModuleSemantics < T::Struct
+          extend T::Sig
+          SemanticsCache = T.type_alias { T::Hash[String, FileSemantics] }
+          const :mod, Module
+          const :semantics, SemanticsCache, factory: -> { {} }
+
+          sig { params(unbound_method: UnboundMethod).returns(FileSemantics) }
+          def file_semantics(unbound_method)
+            path, = T.must(unbound_method.source_location)
```

Comment:
`source_location` may not be non-nil, or may not point to real paths. That could happen with advanced forms of code generation or metaprogramming.